### PR TITLE
Update to Grid docs

### PR DESCRIPTION
**Pull request by sarjon**

### CODE REVIEW

**Comment by sarjon** [comment]
ping @mickaelandrieu for review

File: src/content/1.7/development/components/grid-component.md [hunk 1]
```diff
@@ -7,229 +7,404 @@ weight: 2
...
+
+{{% notice note %}}
+It is _recommended_ to keep your name translatable. To make that easy
+`AbstractGridDefinitionFactory` provides access to translator via `trans()` method.
+{{% /notice %}}
+
+And finally register your Grid definition factory as a service.
+
+```yaml
+prestashop.core.grid.definition.factory.products_definition_factory:
```
Comment (mickaelandrieu) [inline src/content/1.7/development/components/grid-component.md]:
```diff
- prestashop.core.grid.definition.factory.products_definition_factory
+ prestashop.core.grid.definition.factory.product_grid_definition_factory
```

File: src/content/1.7/development/components/grid-component.md [hunk 2]
```diff
@@ -7,229 +7,404 @@ weight: 2
...
+    class: 'PrestaShop\PrestaShop\Core\Grid\Definition\Factory\ProductGridDefinitionFactory'
+    parent: 'prestashop.core.grid.definition.factory.abstract_grid_definition'
+    public: true
+``` 
+
+Most of the time you wont be creating Grid Definition by yourself but delegating this task to other services.
+But in case you need to create Grid Definition by hand, here's how you can do that.
+
+```php
+$productsGridDefinitionFactory = $container->get('prestashop.core.grid.definition.factory.products_definition_factory');
```
Comment (mickaelandrieu) [inline src/content/1.7/development/components/grid-component.md]: service name needs to be updated

File: src/content/1.7/development/components/grid-component.md [hunk 3]
```diff
@@ -7,229 +7,404 @@ weight: 2
...
+## Search Criteria
 
-You can rely on the implementation provided in the component `GridDataFactory` or create your owns if you have specific needs. 
+In Grid component Search Criteria is used for Grid's data sorting, paginating & filtering. Search Criteria
+can be loaded from database, URL query or anywhere else. 
 
-## Grid Query Builder
+{{% notice info %}}
+Grid component itself does not manage Search Criteria but instead it provides interface for it.
+In PrestaShop [Filters component](#) is used to resolve Search Criteria for Grid.
```
Comment (mickaelandrieu) [inline src/content/1.7/development/components/grid-component.md]: I need time to document it, would you mind to create an issue so I won't forget?

File: src/content/1.7/development/components/grid-component.md [hunk 4]
```diff
@@ -7,229 +7,404 @@ weight: 2
...
+And finally register your Grid definition factory as a service.
+
+```yaml
+prestashop.core.grid.definition.factory.products_definition_factory:
+    class: 'PrestaShop\PrestaShop\Core\Grid\Definition\Factory\ProductGridDefinitionFactory'
+    parent: 'prestashop.core.grid.definition.factory.abstract_grid_definition'
+    public: true
+``` 
+
+Most of the time you wont be creating Grid Definition by yourself but delegating this task to other services.
```
Comment (mickaelandrieu) [inline src/content/1.7/development/components/grid-component.md]:
```diff
- wont
+ won't
```

File: src/content/1.7/development/components/grid-component.md [hunk 5]
```diff
@@ -7,229 +7,404 @@ weight: 2
...
+);
+```
+
+## Grid Data
+
+The final part of Grid component is data. Grid Data is stored in `PrestaShop\PrestaShop\Core\Grid\Data\GridData`. 
+
+### Creating Grid Data
+
+Grid component does not create Grid Data directly but instead relays on `PrestaShop\PrestaShop\Core\Grid\Data\Factory\GridDataFactoryInterface`.
```
Comment (mickaelandrieu) [inline src/content/1.7/development/components/grid-component.md]:
```diff
- relays
- relies?
```

ping @Quetzacoalt91 as my english is bad

File: src/content/1.7/development/components/grid-component.md [hunk 6]
```diff
@@ -7,229 +7,404 @@ weight: 2
...
 
-### Once built, everything is frozen
+It is possible to include provided template and modify some parts of it or you can create your own template to render
+Grid. 
 
-You will notice that almost every object build is *final* and have no setters: we avoid every chance to break the Grid component once the build of the Grid have been done and validated.
+## Learn more
 
-But we got you covered, you can alter almost everything *before* the Grid is presented to the view thanks to the available hooks:
+* [How to work with Bulk actions](#)
```
Comment (mickaelandrieu) [inline src/content/1.7/development/components/grid-component.md]: If you add these links, you need to complete them :angel:

File: src/content/1.7/development/components/grid-component.md [hunk 7]
```diff
@@ -7,229 +7,404 @@ weight: 2
...
+* [How to create custom Grid Action](#)
+* [How to create custom Column](#)
+* [Columns reference](#)
+* [Bulk Actions reference](#)
+* [Grid Actions reference](#)
 
+## Summary as a schema
 
-{{< figure src="../img/grid_workflow_hooks.png" title="Grid Workflow" >}}
+{{< figure src="../img/grid-build-schema.png" title="Grid schema" >}}
```
Comment (mickaelandrieu) [inline src/content/1.7/development/components/grid-component.md]: except if I've misread it, you have removed 1 schema: please let them both, as they have not the same utility: understand how it works vs describe the hooks.

File: src/content/1.7/development/components/grid-component.md (hunk 6 above)
Comment (sarjon) [reply src/content/1.7/development/components/grid-component.md]: can I add _Comming soon_ and make different PR's for that?

File: src/content/1.7/development/components/grid-component.md (hunk 7 above)
Comment (sarjon) [reply src/content/1.7/development/components/grid-component.md]: i guess you are right, but is it a correct place for it? I would prefer adding diagram with hooks to "How to modify Grid from modules" chapter as it's relevant to it, wdyt?

File: src/content/1.7/development/components/grid-component.md [hunk 8]
```diff
@@ -7,229 +7,404 @@ weight: 2
...
+{{% /notice %}}
 
-This is the related interface:
+### Creating Search Criteria
+
+Even though most of the time Search Criteria will be created using [Filters component](#), you can still
+create it by yourself. Grid provides simple implementation for it.
+
+```php
+use PrestaShop\PrestaShop\Core\Grid\Search\SearchCriteria;
```
Comment (jolelievre) [inline src/content/1.7/development/components/grid-component.md]: weird thing, why is `SearchCriteriaInterface` the only one in `PrestaShop\PrestaShop\Core\Grid\Search` namespace when all filters that implement it are in the `PrestaShop\PrestaShop\Core\Search` namespace, shoudn't the Interface move?
either it or the Filters classes

File: src/content/1.7/development/components/grid-component.md [hunk 9]
```diff
@@ -7,229 +7,404 @@ weight: 2
...
+
+$filters = [
+    'id_product' => 4,
+    'name' => 'mug',
+];
+
+$searchCriteria = new SearchCriteria(
```
Comment (jolelievre) [inline src/content/1.7/development/components/grid-component.md]: SearchCriteria class does not exist (only the Interface), should we add it?

File: src/content/1.7/development/components/grid-component.md [hunk 10]
```diff
@@ -7,229 +7,404 @@ weight: 2
...
+                continue;
+            }
+
+            $qb->andWhere("$filterName LIKE :$filterName");
+            $qb->setParameter($filterName, '%'.$filterValue.'%');
+        }
+
+        return $qb;
+    }
+
```
Comment (jolelievre) [inline src/content/1.7/development/components/grid-component.md]: `getCountQueryBuilder` should integrate the filters clauses, the aim is not to have the total count but the contextual count (with filters)
otherwise how do you paginate when you use a filter? only the `setMaxResults` `setOffset` part should be restricted to `getSearchQueryBuilder`

File: src/content/1.7/development/components/grid-component.md [hunk 11]
```diff
@@ -7,229 +7,404 @@ weight: 2
...
+    class: 'PrestaShop\PrestaShop\Core\Grid\Data\Factory\DoctrineGridDataFactory'
+    arguments:
+        - '@prestashop.core.grid.query.product_query_builder' # service id of our query builder
+        - '@prestashop.core.hook.dispatcher' # every doctrine query builder needs hook dispatcher
+        - 'products' # this should match your grid id, in our case it's "products"
+```
+
+And that's it! Now we can use our Grid Data factory together with Search Criteria to get sorted, paginated and 
+filtered data for our Grid.
+
```
Comment (jolelievre) [inline src/content/1.7/development/components/grid-component.md]: also an example with DoctrineSearchCriteriaApplicatorInterface might be nice? wdyt?
or should it rather be in the "Lear more" section?

File: src/content/1.7/development/components/grid-component.md [hunk 12]
```diff
@@ -7,229 +7,404 @@ weight: 2
...
      */
-    public function getCountQueryBuilder(SearchCriteriaInterface $searchCriteria = null);
+    public function __construct(Connection $connection, $dbPrefix, $contextLangId, $contextShopId)
+    {
+        parent::__construct($connection, $dbPrefix);
 
-```
+        $this->contextLangId = $contextLangId;
+        $this->contextShopId = $contextShopId;
+    }
```
Comment (jolelievre) [inline src/content/1.7/development/components/grid-component.md]: I love that you presented an example with language and shop context!!
which makes me think, maybe we could add an abstract class which does that?

File: src/content/1.7/development/components/grid-component.md (hunk 3 above)
Comment (jolelievre) [reply src/content/1.7/development/components/grid-component.md]: Yeah I noticed this part was missing, as well as all the "Learn more" links
shouldn't we add "Coming soon" message? Are these pages supposed to be separate pages or sub sections of this one?
I wrote a little section about this in my PR which could be integrated until the full page is written
https://github.com/PrestaShop/docs/pull/134/files#diff-a04004808f283fab9e54e5fb0b03e318R238

File: src/content/1.7/development/components/grid-component.md (hunk 10 above)
Comment (sarjon) [reply src/content/1.7/development/components/grid-component.md]: I guess you are right, it should integrate filters. :+1:

File: src/content/1.7/development/components/grid-component.md (hunk 11 above)
Comment (sarjon) [reply src/content/1.7/development/components/grid-component.md]: i dont know really regarding applicator, as it is not actually "fully completed" yet. :) @mickaelandrieu wdyt?

File: src/content/1.7/development/components/grid-component.md (hunk 12 above)
Comment (PierreRambaud) [reply src/content/1.7/development/components/grid-component.md]: I love this idea, this will be help many people :)

File: src/content/1.7/development/components/grid-component.md (hunk 8 above)
Comment (jolelievre) [reply src/content/1.7/development/components/grid-component.md]: ok I get the idea But then if Filters are only used by the Grid, which would appear so since they implement the `SearchCriteriaInterface ` shouldn't they be in the Grid namespace as well

File: src/content/1.7/development/components/grid-component.md (hunk 7 above)
Comment (mickaelandrieu) [reply src/content/1.7/development/components/grid-component.md]: make sense :+1:

**Comment by mickaelandrieu** [comment]
I'm not done in the rebase *yet*, not really sure of the right structure of this part.

**Comment by mickaelandrieu** [comment]
Hello @sarjon, would you mind to rebase your pull request?

Contents are organization of the main Grid page has changed a lot, I can't figure out what do I need to keep, adapt and remove :thinking: 

Thanks!

File: src/content/1.7/development/components/_index.md [hunk 13]
```diff
@@ -7,5 +7,7 @@ weight: 50
 # Components
 
 * [Grid Component][grid-component]
+* [Faceted Search Component][faceted-search-component]
```
Comment (sarjon) [inline src/content/1.7/development/components/_index.md]: it would be nice to separate FO and BO components, maybe with labels or separate sections, like FO components list and BO components list. :)

**Comment by sarjon** [comment]
@mickaelandrieu rebase done, but needs to be double checked. 

by the way, when reading docs (and not only), i've noticed that "Grid actions" are not the best name (we already discussed it with @matks but didnt come up with new name :disappointed:). I think that "Toolbar buttons/actions" would be better and easier to understand for PS developers since it's the same name as in legacy.

another thing is related to `DataFactory`, i would prefer `DataProvider` as its much easier to understand and gives you clear idea of what it does imo.

so, wdyt?

File: src/content/1.7/development/components/grid/_index.md [hunk 14]
```diff
@@ -1,324 +1,410 @@
...
+Even though most of the time Search Criteria will be created using [Filters component](#), you can still
+create it by yourself. Grid provides simple implementation for it.
 
-```yaml
-# In src/PrestaShopBundle/Resources/config/services/core/grid.yml
-parameters:
-  prestashop.core.grid.data.factory.doctrine_grid_data_factory: PrestaShop\PrestaShop\Core\Grid\Data\Factory\DoctrineGridDataFactory
+```php
+use PrestaShop\PrestaShop\Core\Grid\Search\SearchCriteria;
+
```
Comment (jolelievre) [inline src/content/1.7/development/components/grid/_index.md]: this class does not exist in 1.7.5, should we specify the minimum version to use this class?

File: src/content/1.7/development/components/grid/_index.md [hunk 15]
```diff
@@ -1,324 +1,410 @@
...
+$searchCriteria->getOrderWay(); // "asc"
+$searchCriteria->getOffset();   // 0
+$searchCriteria->getLimit();    // 10
+```
 
-services:
-  _defaults:
-    public: true
+When creating Search Criteria you can skip some or all it's data. You can leave default `null` values for order way
+and order by to disable sorting or offset and limit to disable pagination.
```
Comment (jolelievre) [inline src/content/1.7/development/components/grid/_index.md]: so `offset` has to be nullified to disable pagination? I would have thought nullifying `limit` is more logical

File: src/content/1.7/development/components/grid/_index.md [hunk 16]
```diff
@@ -1,324 +1,410 @@
...
+        - "@=service('prestashop.adapter.legacy.context').getContext().shop.id"
+    public: true
+    
+# Configure Grid Data factory to use query builder that we registered above
+prestashop.core.grid.data.factory.product_data_factory:
+    class: 'PrestaShop\PrestaShop\Core\Grid\Data\Factory\DoctrineGridDataFactory'
+    arguments:
+        - '@prestashop.core.grid.query.product_query_builder' # service id of our query builder
+        - '@prestashop.core.hook.dispatcher' # every doctrine query builder needs hook dispatcher
+        - 'products' # this should match your grid id, in our case it's "products"
```
Comment (jolelievre) [inline src/content/1.7/development/components/grid/_index.md]: `DoctrineGridDataFactory` needs the dispatcher, not the Query builder

**Comment by mickaelandrieu** [comment]
> another thing is related to DataFactory, i would prefer DataProvider as its much easier to understand and gives you clear idea of what it does imo.

It's not a provider, but a factory and this make the architecture of the component inconsistent, so I disagree. It doesn't mean we can't invest in docs to better explain what is a factory to PrestaShop developers.

> this class does not exist in 1.7.5, should we specify the minimum version to use this class?

No, the example should be updated using type hint on the interface instead

> by the way, when reading docs (and not only), i've noticed that "Grid actions" are not the best name (we already discussed it with @matks but didnt come up with new name disappointed). I think that "Toolbar buttons/actions" would be better and easier to understand for PS developers since it's the same name as in legacy.

It's too late for that: we are so close to relate 1.7.5.0 so the QA team won't accept such big changes on API which are not bug fixes and we will not create a BC on 1.7.6 => needs to be planned for 1.8.x

**Comment by sarjon** [comment]
i wont be available much this week. :/

File: src/content/1.7/development/components/grid/_index.md (hunk 15 above)
Comment (sarjon) [reply src/content/1.7/development/components/grid/_index.md]: if you want to disable pagination, you have to nullify both `limit` and `offset`, you can choose to disable one of them only (e.g. disable offset, but keep limit).

**Comment by matks** [comment]
ping @jolelievre 😉

File: src/content/1.7/development/components/grid/_index.md (hunk 15 above)
Comment (jolelievre) [reply src/content/1.7/development/components/grid/_index.md]: I would rephrase this sentence, it's longer but clearer:

When creating Search Criteria you can skip some or all of its data. If you set **both** `orderWay` and `orderBy` to `null` it will disable sorting.  If you set **both** `offset` and `limit` to `null` it will disable pagination.

**Comment by jolelievre** [comment]
@sarjon only a last small change suggested, the rest si good for me as well
